### ISSUE TITLE
Can't compile Scala 2.12 to javascript

### ISSUE THREAD

**Issue by scf37** [konsoletyper/teavm#231]
`[ERROR] Substitutor for bootstrap method java.lang.invoke.LambdaMetafactory.altMetafactory(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;[Ljava/lang/Object;)Ljava/lang/invoke/CallSite; was not found`

I'm using maven archetype project  for teavm with scala compiler added. Type `mvn package` to reproduce.
[teavm.zip](https://github.com/konsoletyper/teavm/files/606049/teavm.zip)

**Comment by konsoletyper** [comment]
Looks like Scala 2.12 compiler produces `INVOKEDYNAMIC` instructions with `LambdaMetafactory.altMetafactory` as bootstrap method. This method is unsupported in TeaVM. Does Scala compiler provide an option to compile lambdas to classes instead of INVOKEDYNAMIC? If not, try to setup retrolambda. It's a quick workaround to make things just work.

I believe it's easy to support `altMetafactory`. It should contain just few things over `metafactory` method that is already implemented.

**Comment by scf37** [comment]
It is Scala 2.12 main feature - compilation to lambdas instead of inner classes. I was hoping teavm produces smaller javascript for that case.

**Comment by konsoletyper** [comment]
TeaVM does support `invokedynamic`, but does not support serializable lambdas (which are `LambdaMetafactory.altMetafactory`). TeaVM won't generate less code for `invokedynamic` lambdas, it converts `invokedynamic` to a class.

**Comment by konsoletyper** [comment]
I think it's possible to mark the generated class so that if further eliminated and replaced by the code that constructs lambda in run-time. But it's impossible to avoid this temporary class due to need of dependency analysis.

**Comment by scf37** [comment]
FYI, js size comparsion for this example between ScalaJS and teavm:

 Project | Minified | Gzipped
------------ | --------- | --------
 TeaVM |  270K    | 64K
 ScalaJS | 153K   | 38K

So, while using jvm libraries as-is is very appealing...

**Comment by konsoletyper** [comment]
Of course, TeaVM produces some overhead compared to ScalaJS, which has a lot of more information and hence can sometimes generate less code. However, TeaVM's advantage is not *less* produced code compared to ScalaJS. Its main advantage is ability to compile mixed Scala/Java projects.

I don't think that lambdas are the main source of overhead. There are many things that can be improved, like stripping unnecessary `return` statements, remove unused class metadata, etc.

**Comment by scf37** [comment]
Looking at generated code, teavm compiles Scala 'idiomatic' stuff to lots of primitive, and often identical functions:
```
function Xj($t,a){$t.hQ=a;return;}
function U1($t,a){$t.mH=a;return;}
function GC(){E.call(this);}
function FB(){E.call(this);}
function AB(){OG.call(this);}
function CD(){E.call(this);}
.....
```

For some reason, `optimizationLevel>FULL</optimizationLevel>` does not help with inlining. Just to let you know.

**Comment by konsoletyper** [comment]
I can't tell anything. I should examine the generated code myself. How should I change build configuration to make your project compilable by TeaVM?

The probable reason for repeated code: it's just repeated in bytecode. TeaVM does not try to eliminate repeated code to make it more monomorphic.

What about `optimizationLevel`? Do you use latest TeaVM built from source code? How did you determine that inlining does not happen? Did you try to examine the code with minification disabled?

**Comment by scf37** [comment]
To make it compilable, replace 2.12.0 with 2.11.8 over pom.xml

I think I'll leave TeaVM for now. I was looking for toolkit for public web - that produces small and fast javascript for mobile devices with restricted cpu power and network bandwidth. TeaVM seems to fill another niche - intranet applications and games.
